### ISSUE TITLE
Does this action no more place a "git-ftp.lck" file?

### ISSUE THREAD

**Issue by xannasavin** [SamKirkland/FTP-Deploy-Action#242]
**Bug Description**

In the past, I noticed that the deploy action was placing a "git-ftp.lck" file in the folder of deployment.
I made use of this and built some automatisms, that put pages into maintenance mode when this file is present.
However, today during deployment, I noticed that no such file was placed anymore... 

Was something changed, do I miss something?

**My Action Config**
```yaml
# https://github.com/marketplace/actions/ftp-deploy
# 31.05.2021

name: 🚀 Deploy website on push

# Controls when the action will run. 
on:
  # Triggers the workflow on push or pull request events but only for the master branch
  push:
    branches: [ main ]

  # Allows you to run this workflow manually from the Actions tab
  workflow_dispatch:

# A workflow run is made up of one or more jobs that can run sequentially or in parallel
jobs:
  web-deploy:
    name: 🎉 Deploy
    runs-on: ubuntu-latest
    steps:
    - name: 🚚 Get latest code
      uses: actions/checkout@v2.3.2

    
    - name: 📂 Sync files
      uses: SamKirkland/FTP-Deploy-Action@4.0.0
      with:
        server: ${{ secrets.FTP_SERVER }}
        username: ${{ secrets.FTP_USER }}
        password: ${{ secrets.FTP_PASSWORD }}
        exclude: .git*
          - .git*/**
          -  **/.git*/**
          - .github/*
          - .gitignore
          - */.gitignore
          - */.gitkeep
          - .gitlab-ci.yml
          - readme.md
          - *.md
          - /logs/*  
          - /upgrade/*
          - /uploads/*
          - /languages/*
          - /updraft/*
          - /cache/*  
          - *.scss
          - */.idea/
          - */dev-src/
          - prepros.config
          - *.map
```

**Comment by xannasavin** [comment]
it seems this does not happen since something longer. maybe with the new update - can't tell, the time I built those mechanisms is quite some time ago and I just noticed today ><

Is there any way I can tell the action to place such file in the beginning and remove it afterwards? Would be really helpfull...

**Comment by SamKirkland** [comment]
v4+ no longer uses that file.
v4+ creates a `.ftp-deploy-sync-state.json` file, but it is always present. The file is updated after every successful deployment.

One option would be to have a maintenance page and redirect to it when the deployment starts, then remove the redirect when the deployment ends. For example if you are using Apache to host your website you could use a `htaccess` file to redirect users to maintenance page, and remove the file when the deployment is complete.

I'm not sure what tooling you were using to do this previously. But if you wanted to use github actions you could use a step before `ftp-deploy-action` to deploy the htaccess file, and another step after the deployment to remove the `htaccess` file.

I wrote some code below, the important bit is `state-name` - we need to use a unique state name for the maintenance deploy, state-name is used so the script knows which files exist on the server.

Example: Note the following is untested example code only
```yaml
jobs:
  web-deploy:
    name: 🎉 Deploy
    runs-on: ubuntu-latest
    steps:
    - name: 🚚 Get latest code
      uses: actions/checkout@v2.3.2

    - name: 👷 Maintenance mode on
      uses: SamKirkland/FTP-Deploy-Action@4.0.0
      with:
        server: ${{ secrets.FTP_SERVER }}
        username: ${{ secrets.FTP_USER }}
        password: ${{ secrets.FTP_PASSWORD }}
        state-name: .ftp-deploy-sync-state--maintenance-mode.json
        exclude: EXCLUDE ALL FILES EXCEPT THE HTACCESS FILE AND MAINTENCE PAGE

    - name: 📂 Sync files
      uses: SamKirkland/FTP-Deploy-Action@4.0.0
      with:
        server: ${{ secrets.FTP_SERVER }}
        username: ${{ secrets.FTP_USER }}
        password: ${{ secrets.FTP_PASSWORD }}
        exclude: EXCLUDE HTACCESS FILE, BECAUSE THIS IS USING A DIFFERENT STATE-NAME IT WONT KNOW THE MAINTENCE PAGE/HTACCESS FILE EXIST... SO THEY WONT BE REMOVED

    - name: 👷 Maintenance mode off
      uses: SamKirkland/FTP-Deploy-Action@4.0.0
      with:
        server: ${{ secrets.FTP_SERVER }}
        username: ${{ secrets.FTP_USER }}
        password: ${{ secrets.FTP_PASSWORD }}
        state-name: .ftp-deploy-sync-state--maintenance-mode.json
        exclude: EXCLUDE ALL FILES SO THE HTACCESS FILE IS DELETED
```


I'm going to close this out. If you get something working please post back here with your htaccess file and deployment yaml and I can add it as an example!

**Comment by xannasavin** [comment]
That's an awesome hint, thanks a lot. I will play around with it and come back.

So if I understand correctly, the file I want to place on the server to trigger maintenance mode, needs to be in the repo and will be placed with the first action and removed with the last.

Since my htaccess is generated and changed on serverside by WordPress, that approach won't work.
But since I just need any file on the server to look for with the WordPress plugin I created, I can just place a file and look for that - so that should work.

I'll come back when I figured it out :)

**Comment by SamKirkland** [comment]
Yep your understanding is correct.

You don't have to modify/delete the htaccess file, you can actually check for existence of the file with htaccess.
See https://stackoverflow.com/a/20202483

**Comment by xannasavin** [comment]
Quick question: how can I simply exclude all files in the last  "Maintenance Mode Off" job? Can I simply put an * after the exclude? Or do I have to add all files and folders existing? Like
`exclude: *`

For the first "Maintenance Mode On" job a similar question: is it possible to put a * to exclude and then include files with e.g. ! like it is possible in the .gitignore? so e.g.
```
exclude: *
- !maintenance-lock-file.lck
```

Sorry, I'm not really into this and couldn't find anything helpfull to this question.

Thank you for your answer!

**Comment by xannasavin** [comment]
@SamKirkland a quick hint on the above would be much appreciated! thank you

**Comment by SamKirkland** [comment]
@xannasavin I took a look at this over lunch and got it working on one of my sites.

See my updated comment above for example code. I've verified this works on my apache based websites. Please test it out and let me know, then I can add it to the examples

**Comment by xannasavin** [comment]
@SamKirkland thanks a bunch! that "workaround" with the folder works really great!
As said, I'm using a WP Plugin for the job of the maintenance mode, so it looks slightly different on my side, but like this it works!
It places and removes the "maintenance.lck" file inside /maintlck/

However, I noticed one strange thing: I updated the workflow file to use 4.2.0 and now the .git folders are all synced, though I excluded them. With 4.0.0 this seems not to happen. I changed now the "Sync files" back to 4.0.0.
I can open a new issue if needed...

`# https://github.com/marketplace/actions/ftp-deploy
# 25.11.2021

name: 🚀 Deploy website on push

# Controls when the action will run.
on:
  # Triggers the workflow on push or pull request events but only for the master branch
  push:
    branches: [ main ]
    # pull_request:
    # branches: [ master ]

  # Allows you to run this workflow manually from the Actions tab
  workflow_dispatch:

# A workflow run is made up of one or more jobs that can run sequentially or in parallel
jobs:
  web-deploy:
    name: 🎉 Deploy
    runs-on: ubuntu-latest
    steps:
      - name: 🚚 Get latest code
        uses: actions/checkout@v2.3.2
          #with:
          #fetch-depth: 100
          # get the last XXX commits from history
          # Make sure this number is AT LEAST 2, that's because we need to see what has changed.

      - name: 👷 Maintenance mode on
        uses: SamKirkland/FTP-Deploy-Action@4.2.0
        with:
          server: ${{ secrets.FTP_SERVER }}
          username: ${{ secrets.FTP_USER }}
          password: ${{ secrets.FTP_PASSWORD }}
          state-name: .ftp-deploy-sync-state--maintenance-mode.json
          local-dir: ./maintlck/
          server-dir: ./maintlck/

      - name: 📂 Sync files
        uses: SamKirkland/FTP-Deploy-Action@4.0.0
        with:
          server: ${{ secrets.FTP_SERVER }}
          username: ${{ secrets.FTP_USER }}
          password: ${{ secrets.FTP_PASSWORD }}
          exclude: .git*
            - .git*/**
            -  **/.git*/**
            - .github/*
            - .gitignore
            - */.gitignore
            - */.gitkeep
            - .gitlab-ci.yml
            - readme.md
            - *.md
            - /logs/*
            - /upgrade/*
            - /uploads/*
            - /languages/*
            - /updraft/*
            - /cache/*
            - /backups-dup-lite/*
            - /envato-backups/
            - /mu-plugins/
            - /tmp/
            - *.scss
            - */.idea/
            - */dev-src/
            - prepros.config
            - *.map
            - */scss/
            - /maintlck/

      - name: 👷 Maintenance mode off
        uses: SamKirkland/FTP-Deploy-Action@4.2.0
        with:
          server: ${{ secrets.FTP_SERVER }}
          username: ${{ secrets.FTP_USER }}
          password: ${{ secrets.FTP_PASSWORD }}
          state-name: .ftp-deploy-sync-state--maintenance-mode.json
          local-dir: ./maintlck/
          server-dir: ./maintlck/
          exclude: |
            maintenance.lck`

**Comment by SamKirkland** [comment]
Glad it's working!
The exclude format in v4.2.0 was changed, see the migration guide for more info.

**Comment by xannasavin** [comment]
ah got it! 
Thanks a lot again for your help, that's really awesome!

**Comment by xannasavin** [comment]
Need to reopen this. It looks like the lock file is no more removed at the end of the workflow. I updated to 4.3.1 - was there a change to this?

`name: 🚀 Deploy website on push

# Controls when the action will run.
on:
  # Triggers the workflow on push or pull request events but only for the master branch
  push:
    branches: [ main ]
    # pull_request:
    # branches: [ master ]

  # Allows you to run this workflow manually from the Actions tab
  workflow_dispatch:

# A workflow run is made up of one or more jobs that can run sequentially or in parallel
jobs:
  web-deploy:
    name: 🎉 Deploy
    runs-on: ubuntu-latest
    steps:
      - name: 🚚 Get latest code
        uses: actions/checkout@v2
          #with:
          #fetch-depth: 100
          # get the last XXX commits from history
          # Make sure this number is AT LEAST 2, that's because we need to see what has changed.

      - name: 👷 Maintenance mode on
        uses: SamKirkland/FTP-Deploy-Action@4.3.1
        with:
          server: ${{ secrets.FTP_SERVER }}
          username: ${{ secrets.FTP_USER }}
          password: ${{ secrets.FTP_PASSWORD }}
          state-name: .ftp-deploy-sync-state--maintenance-mode.json
          local-dir: ./maintlck/
          server-dir: ./maintlck/

      - name: 📂 Sync files
        uses: SamKirkland/FTP-Deploy-Action@4.3.1
        with:
          server: ${{ secrets.FTP_SERVER }}
          username: ${{ secrets.FTP_USER }}
          password: ${{ secrets.FTP_PASSWORD }}
          exclude: |
            **/.git*
            **/.git*/**
            .git*
            .git*/**
            .github/*
            .gitignore
            */.gitignore
            */.gitkeep
            .gitlab-ci.yml
            readme.md
            *.md
            /logs/*
            /upgrade/*
            /uploads/*
            /languages/*
            /updraft/*
            /cache/*
            /backups-dup-lite/*
            /envato-backups/
            /mu-plugins/
            /tmp/
            *.scss
            */.idea/
            */dev-src/
            prepros.config
            *.map
            */scss/
            /maintlck/
      - name: 👷 Maintenance mode off
        uses: SamKirkland/FTP-Deploy-Action@4.3.1
        with:
          server: ${{ secrets.FTP_SERVER }}
          username: ${{ secrets.FTP_USER }}
          password: ${{ secrets.FTP_PASSWORD }}
          state-name: .ftp-deploy-sync-state--maintenance-mode.json
          local-dir: ./maintlck/
          server-dir: ./maintlck/
          exclude: |
            maintenance.lck`

**Comment by xannasavin** [comment]
Hi @SamKirkland are you able to help with this?
